Cython/Compiler/ModuleNode.py

```diff
@@ -2542,151 +2516,178 @@ def generate_import_star(self, env, code):
-            Naming.modulestateglobal_cname,
-            Naming.modulestate_cname,
-            Naming.pymoduledef_cname))
-        module_state.putln('')
-        module_state.putln('#define %s (PyState_FindModule(&%s))' % (
-            env.module_cname,
-            Naming.pymoduledef_cname))
-        module_state.putln("#endif")
-        module_state_defines.putln("#endif")
-        module_state_clear.putln("return 0;")
-        module_state_clear.putln("}")
-        module_state_clear.putln("#endif")
-        module_state_traverse.putln("return 0;")
-        module_state_traverse.putln("}")
-        module_state_traverse.putln("#endif")
-
-    def generate_module_state_defines(self, env, code):
         code.putln("#if CYTHON_USE_MODULE_STATE")
-        code.putln('#define %s %s->%s' % (
-            Naming.builtins_cname,
-            Naming.modulestateglobal_cname,
-            Naming.builtins_cname))
-        code.putln('#define %s %s->%s' % (
-            Naming.cython_runtime_cname,
-            Naming.modulestateglobal_cname,
-            Naming.cython_runtime_cname))
-        code.putln('#define %s %s->%s' % (
-            Naming.empty_tuple,
-            Naming.modulestateglobal_cname,
-            Naming.empty_tuple))
-        code.putln('#define %s %s->%s' % (
-            Naming.empty_bytes,
-            Naming.modulestateglobal_cname,
-            Naming.empty_bytes))
-        code.putln('#define %s %s->%s' % (
-            Naming.empty_unicode,
-            Naming.modulestateglobal_cname,
-            Naming.empty_unicode))
-        if Options.pre_import is not None:
-            code.putln('#define %s %s->%s' % (
-                Naming.preimport_cname,
-                Naming.modulestateglobal_cname,
-                Naming.preimport_cname))
-        code.putln('#ifdef __Pyx_CyFunction_USED')
-        code.putln('#define %s %s->%s' % (
-            Naming.cyfunction_type_cname,
-            Naming.modulestateglobal_cname,
-            Naming.cyfunction_type_cname))
+
+    def generate_module_state_end(self, env, code):
+        code.putln('#endif /* CYTHON_USE_MODULE_STATE */')
+        code.putln('} %s;' % Naming.cglobals_type_cname)
+        code.putln('')
+        code.putln('#if !CYTHON_USE_MODULE_STATE')
+        code.putln('static %s %s;' % (Naming.cglobals_type_cname, Naming.cglobals_cname))
+        code.putln('#define __Pyx_CGlobal(ident) (%s.ident)' % Naming.cglobals_cname)
+        code.putln('static PyObject *%s;' % env.module_cname)
+        code.putln('#else')
+        code.putln('/* Module def forward declaration */')
+        code.putln('#ifdef __cplusplus')
+        code.putln('namespace {')
+        code.putln('extern struct PyModuleDef %s;' % Naming.pymoduledef_cname)
+        code.putln('} /* anonymous namespace */')
+        code.putln('#else')
+        code.putln('static struct PyModuleDef %s;' % Naming.pymoduledef_cname)
         code.putln('#endif')
```

Comment:
I also always wondered why this C++ special case was introduced. Any idea? Why should this be necessary?

Reply:
Yes, turns out in C you can forward-declare a static global variable (it's technically not a forward declaration, but just two definitions that get merged by the linker). While in C++ you can only forward declare via `extern` which would mean the variable has no longer private linkage. Hence this strange workaround puts the variable into an anonymous namespace so it is private again...

Reply:
That would make an excellent comment in the code, I think. :)

Reply:
Turns out this can be fixed in a simpler manner by simply moving the `PyModuleDef` instance above the module state so we no longer need a forward declaration.